src/laserscan_to_pointcloud_node.cpp

```diff
@@ -38,31 +38,108 @@
+#include "tf2_sensor_msgs/tf2_sensor_msgs.h"
+#include "tf2_ros/create_timer_ros.h"
+
+namespace pointcloud_to_laserscan
+{
 
-int main(int argc, char** argv)
+LaserScanToPointCloudNode::LaserScanToPointCloudNode(const rclcpp::NodeOptions & options)
+: rclcpp::Node("laserscan_to_pointcloud", options)
 {
-  ros::init(argc, argv, "laserscan_to_pointcloud_node");
-  ros::NodeHandle private_nh("~");
-  int concurrency_level = private_nh.param("concurrency_level", 0);
-
-  nodelet::Loader nodelet;
-  nodelet::M_string remap(ros::names::getRemappings());
-  nodelet::V_string nargv;
-  std::string nodelet_name = ros::this_node::getName();
-  nodelet.load(nodelet_name, "pointcloud_to_laserscan/laserscan_to_pointcloud_nodelet", remap, nargv);
-
-  boost::shared_ptr<ros::MultiThreadedSpinner> spinner;
-  if (concurrency_level)
-  {
-    spinner.reset(new ros::MultiThreadedSpinner(static_cast<uint32_t>(concurrency_level)));
+  target_frame_ = this->declare_parameter("target_frame", "");
+  tolerance_ = this->declare_parameter("transform_tolerance", 0.01);
+  // TODO(hidmic): adjust default input queue size based on actual concurrency levels
+  // achievable by the associated executor
+  input_queue_size_ = this->declare_parameter(
+    "queue_size", static_cast<int>(std::thread::hardware_concurrency()));
+
+  pub_ = this->create_publisher<sensor_msgs::msg::PointCloud2>("cloud", rclcpp::SensorDataQoS());
+
+  using std::placeholders::_1;
+  // if pointcloud target frame specified, we need to filter by transform availability
+  if (!target_frame_.empty()) {
+    tf2_.reset(new tf2_ros::Buffer(this->get_clock()));
+    auto timer_interface = std::make_shared<tf2_ros::CreateTimerROS>(
+      this->get_node_base_interface(), this->get_node_timers_interface());
+    tf2_->setCreateTimerInterface(timer_interface);
+    tf2_listener_.reset(new tf2_ros::TransformListener(*tf2_));
+    message_filter_.reset(new MessageFilter(sub_, *tf2_, target_frame_, input_queue_size_,
+      this->get_node_logging_interface(),
+      this->get_node_clock_interface()));
+    message_filter_->registerCallback(std::bind(&LaserScanToPointCloudNode::scanCallback, this,
+      _1));
+  } else {  // otherwise setup direct subscription
+    sub_.registerCallback(std::bind(&LaserScanToPointCloudNode::scanCallback, this, _1));
   }
-  else
-  {
-    spinner.reset(new ros::MultiThreadedSpinner());
+
+  subscription_listener_thread_ = std::thread(
+    std::bind(&LaserScanToPointCloudNode::subscriptionListenerThreadLoop, this));
+}
+
+LaserScanToPointCloudNode::~LaserScanToPointCloudNode()
+{
+  alive_.store(true);
+  subscription_listener_thread_.join();
+}
+
+void LaserScanToPointCloudNode::subscriptionListenerThreadLoop()
+{
+  rclcpp::Context::SharedPtr context = this->get_node_base_interface()->get_context();
+
+  const std::chrono::milliseconds timeout(100);
+  while (rclcpp::ok(context) && alive_.load()) {
+    int subscription_count = pub_->get_subscription_count();
```

Comment:
I think this is insufficient, particularly in the case where this is part of an intra-process pipeline.  See https://github.com/ros-drivers/velodyne/blob/ec9adec6a523ab705ee526b06141aa9ad8d43c4f/velodyne_pointcloud/src/conversions/convert.cpp#L150 for an example of how to do this correctly.

Same in `pointcloud_to_laserscan_node.cpp` below.

Reply:
Uh, intuition failed me. I would've expected `get_subscription_count()` to get me the total true count. Anyways, good catch and fixed!